### PR TITLE
feat: [CDS-46920]: support linking of org/account level service/env

### PR DESCRIPTION
**Pull request by hingerabhinav**
## Describe your changes

1. changes at service layer for entities: Service, Env to handle service identifier as well as ref.
2. validation helper methods changes to support refs
3. support services list with services accessible at a certain scope. Eg. Call at project level will have all 3 levels of service.
4. fixed issue with env group validation (null safe) and env group listing. 
5. service step v3 changes
6. infra definition entity reference is supported at org/acc levels 

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41102)

### CODE REVIEW

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/EnvironmentValidationHelper.java` line 26 at r3](https://reviewable.io/reviews/harness/harness-core/41102#-NIcK95e4Eon3LVKdbaJ:-NIcK95e4Eon3LVKdbaK:bdxhq4p) ([raw file](https://github.com/harness/harness-core/blob/a462adce5be47ee4f6f1fd1ce35de7717cdf53c6/120-ng-manager/src/main/java/io/harness/ng/core/EnvironmentValidationHelper.java#L26)):*
> ```Java
>   @Inject private EnvironmentService environmentService;
> 
>   public boolean checkThatEnvExists(@NotEmpty String accountIdentifier, @NotEmpty String orgIdentifier,
> ```

can we replace these `org.hibernate.validator.constraints.NotEmpty` by the ones from javax

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/EnvironmentValidationHelper.java` line 26 at r3](https://reviewable.io/reviews/harness/harness-core/41102#-NIcRMtR0oWXC9NMRw0u:-NIcRMtR0oWXC9NMRw0v:b-hxy0ir) ([raw file](https://github.com/harness/harness-core/blob/a462adce5be47ee4f6f1fd1ce35de7717cdf53c6/120-ng-manager/src/main/java/io/harness/ng/core/EnvironmentValidationHelper.java#L26)):*
> ```Java
>   @Inject private EnvironmentService environmentService;
> 
>   public boolean checkThatEnvExists(@NotEmpty String accountIdentifier, @NotEmpty String orgIdentifier,
> ```

also these constraints are no more valid, ProjectIdentifier/OrgIdentifier can be empty

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/service/ServiceEntityValidationHelper.java` line 26 at r3](https://reviewable.io/reviews/harness/harness-core/41102#-NIcRgGW7OVsIqvKqk9a:-NIcRgGW7OVsIqvKqk9b:b-9kjtzz) ([raw file](https://github.com/harness/harness-core/blob/a462adce5be47ee4f6f1fd1ce35de7717cdf53c6/120-ng-manager/src/main/java/io/harness/ng/core/service/ServiceEntityValidationHelper.java#L26)):*
> ```Java
>   @Inject private ServiceEntityService serviceEntityService;
> 
>   public void checkThatServiceExists(@NotEmpty String accountIdentifier, @NotEmpty String orgIdentifier,
> ```

can we replace these org.hibernate.validator.constraints.NotEmpty by the ones from javax

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/service/ServiceEntityValidationHelper.java` line 26 at r3](https://reviewable.io/reviews/harness/harness-core/41102#-NIcRbIB54tZ2xG-4e-2:-NIcRbIB54tZ2xG-4e-3:b-yeoeb4) ([raw file](https://github.com/harness/harness-core/blob/a462adce5be47ee4f6f1fd1ce35de7717cdf53c6/120-ng-manager/src/main/java/io/harness/ng/core/service/ServiceEntityValidationHelper.java#L26)):*
@NotEmpty
> these constraints are no more valid, ProjectIdentifier/OrgIdentifier can be empty

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java` line 472 at r3](https://reviewable.io/reviews/harness/harness-core/41102#-NIcgj6B1pxmdCehIa7u:-NIcgj6B1pxmdCehIa7v:bqafub) ([raw file](https://github.com/harness/harness-core/blob/a462adce5be47ee4f6f1fd1ce35de7717cdf53c6/125-cd-nextgen/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java#L472)):*
> ```Java
>   private ServicePartResponse executeServicePart(
>       Ambiance ambiance, ServiceStepV3Parameters stepParameters, Map<FreezeEntityType, List<String>> entityMap) {
>     // use serviceRef to get identifier ref
> ```

nit: comment is not needed here

**Review by hingerabhinav — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41102)* status: 21 file(s) to review. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/EnvironmentValidationHelper.java` line 26 at r3](https://reviewable.io/reviews/harness/harness-core/41102#-NIcK95e4Eon3LVKdbaJ:-NIlmYWN-Y86LFRzxXhp:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a462adce5be47ee4f6f1fd1ce35de7717cdf53c6/120-ng-manager/src/main/java/io/harness/ng/core/EnvironmentValidationHelper.java#L26)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can we replace these `org.hibernate.validator.constraints.NotEmpty` by the ones from javax
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/EnvironmentValidationHelper.java` line 26 at r3](https://reviewable.io/reviews/harness/harness-core/41102#-NIcRMtR0oWXC9NMRw0u:-NIlmYv_CuStdGEFeAVy:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a462adce5be47ee4f6f1fd1ce35de7717cdf53c6/120-ng-manager/src/main/java/io/harness/ng/core/EnvironmentValidationHelper.java#L26)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

also these constraints are no more valid, ProjectIdentifier/OrgIdentifier can be empty
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/service/ServiceEntityValidationHelper.java` line 26 at r3](https://reviewable.io/reviews/harness/harness-core/41102#-NIcRgGW7OVsIqvKqk9a:-NIlmaad46qY1yU5T2QS:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a462adce5be47ee4f6f1fd1ce35de7717cdf53c6/120-ng-manager/src/main/java/io/harness/ng/core/service/ServiceEntityValidationHelper.java#L26)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

can we replace these org.hibernate.validator.constraints.NotEmpty by the ones from javax
</blockquote></details>

Done.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/service/ServiceEntityValidationHelper.java` line 26 at r3](https://reviewable.io/reviews/harness/harness-core/41102#-NIcRbIB54tZ2xG-4e-2:-NIlmbBPDQ-lp5ZQrjDE:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/a462adce5be47ee4f6f1fd1ce35de7717cdf53c6/120-ng-manager/src/main/java/io/harness/ng/core/service/ServiceEntityValidationHelper.java#L26)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

@NotEmpty
> these constraints are no more valid, ProjectIdentifier/OrgIdentifier can be empty

</blockquote></details>

Done.
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java` line 442 at r3](https://reviewable.io/reviews/harness/harness-core/41102#-NIcgzly8ugtzYBmdCFk:-NIlmmkaCpMM0p5VF7-t:byffcz8) ([raw file](https://github.com/harness/harness-core/blob/a462adce5be47ee4f6f1fd1ce35de7717cdf53c6/127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java#L442)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

we should try to reduce the number of DB queries here ?
We have 2 options here
If we know that all service identifiers are from same scope, we can just make 1 db query
If we know that all service identifiers might not be from same scope, we can just make 1 db query per scope
</blockquote></details>

Done. This is called from pipelines and with multi service. Added handling for different scopes.

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/ServiceEntityValidationHelper.java [hunk 1]
```diff
@@ -9,24 +9,41 @@
...
-    if (!service.isPresent()) {
-      throw new NotFoundException(String.format("service [%s] not found.", serviceIdentifier));
+  public void checkThatServiceExists(
+      @NotEmpty String accountIdentifier, String orgIdentifier, String projectIdentifier, @NotEmpty String serviceRef) {
+    Optional<ServiceEntity> service;
+    String[] serviceRefSplit = serviceRef.split("\\.", 2);
+    // project level entity or org/account level entity with identifier
+    if (serviceRefSplit.length == 1) {
+      service = serviceEntityService.get(accountIdentifier, orgIdentifier, projectIdentifier, serviceRef, false);
+    } else if (serviceRefSplit.length == 2) {
```
Comment (hingerabhinav) [inline 120-ng-manager/src/main/java/io/harness/ng/core/service/ServiceEntityValidationHelper.java:34]: max is anyways 2

**Review by yogesh-chauhan — DISMISSED** [review]
> Reviewed 2 of 5 files at r1, 2 of 3 files at r2, 1 of 2 files at r3, 8 of 19 files at r4, 4 of 11 files at r5, 25 of 25 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41102)* status: 39 file(s) to review. 1 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41102)* status: 39 file(s) to review. 1 of 2 approvals obtained.

**Comment by tchaurasiya** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java` line 501 at r8](https://reviewable.io/reviews/harness/harness-core/41102#-NJ_k3zCAIoKCSljZhmK:-NJ_k3zCAIoKCSljZhmL:bij6e06) ([raw file](https://github.com/harness/harness-core/blob/ce140c56027aa89ed14df34b61781a11112631cb/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java#L501)):*
> ```Java
>               "Specifies sorting criteria of the list. Like sorting based on the last updated entity, alphabetical sorting in an ascending or descending order")
>       @QueryParam("sort") List<String> sort) {
>     accessControlClient.checkForAccessOrThrow(ResourceScope.of(accountId, orgIdentifier, projectIdentifier),
> ```

will this method be used for org/account level environment group?
In that case we should replace this RBAC check.
Also the description,
(Same goes for listEnvironmentsV2() method.



_Code snippet:_
```Java
"Returns the list of Environments for a Project that are accessible"
```

**Review by tchaurasiya — COMMENTED** [review]
> Reviewed 1 of 19 files at r4, 4 of 25 files at r6, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41102)* status: 33 file(s) to review. 1 of 2 approvals obtained.

**Comment by tchaurasiya** [comment]
*[`127-cd-nextgen-entities/src/main/java/io/harness/ng/core/infrastructure/services/impl/InfrastructureEntityServiceImpl.java` line 150 at r8](https://reviewable.io/reviews/harness/harness-core/41102#-NJ_oiAd6RutDs8MCQGM:-NJ_oiAd6RutDs8MCQGN:b-wexcgn) ([raw file](https://github.com/harness/harness-core/blob/ce140c56027aa89ed14df34b61781a11112631cb/127-cd-nextgen-entities/src/main/java/io/harness/ng/core/infrastructure/services/impl/InfrastructureEntityServiceImpl.java#L150)):*
> ```Java
>   @Override
>   public Optional<InfrastructureEntity> get(
>       String accountId, String orgIdentifier, String projectIdentifier, String environmentIdentifier, String infraRef) {
> ```

should it be envRef?

**Comment by tchaurasiya** [comment]
*[`127-cd-nextgen-entities/src/main/java/io/harness/ng/core/infrastructure/services/impl/InfrastructureEntityServiceImpl.java` line 159 at r8](https://reviewable.io/reviews/harness/harness-core/41102#-NJ_opq8ESe_pq8O3JVW:-NJ_opq8ESe_pq8O3JVX:b-wexcgn) ([raw file](https://github.com/harness/harness-core/blob/ce140c56027aa89ed14df34b61781a11112631cb/127-cd-nextgen-entities/src/main/java/io/harness/ng/core/infrastructure/services/impl/InfrastructureEntityServiceImpl.java#L159)):*
> ```Java
> 
>   private Optional<InfrastructureEntity> getInfrastructureByRef(String accountId, String orgIdentifier,
>       String projectIdentifier, String environmentIdentifier, String infraIdentifier) {
> ```

should it be envRef?

**Comment by tchaurasiya** [comment]
*[`127-cd-nextgen-entities/src/main/java/io/harness/ng/core/infrastructure/services/impl/InfrastructureEntityServiceImpl.java` line 258 at r8](https://reviewable.io/reviews/harness/harness-core/41102#-NJ_qZes1n81yPQdUW8S:-NJ_qV6E7Vu_-m8Qv-yB:b4jxec8) ([raw file](https://github.com/harness/harness-core/blob/ce140c56027aa89ed14df34b61781a11112631cb/127-cd-nextgen-entities/src/main/java/io/harness/ng/core/infrastructure/services/impl/InfrastructureEntityServiceImpl.java#L258)):*
> ```Java
>   public HIterator<InfrastructureEntity> listIterator(String accountId, String orgIdentifier, String projectIdentifier,
>       String envIdentifier, Collection<String> identifiers) {
>     // same scope assumption? NO
> ```

nit: still needed?



_Code quote:_
```Java
/ same scope assumption? NO
```

**Review by hingerabhinav — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41102)* status: 35 file(s) to review. 3. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java` line 501 at r8](https://reviewable.io/reviews/harness/harness-core/41102#-NJ_k3zCAIoKCSljZhmK:-NJe-2kw5gjk6CRy9VbX:b6v3oh0) ([raw file](https://github.com/harness/harness-core/blob/ce140c56027aa89ed14df34b61781a11112631cb/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java#L501)):*
<details><summary><i>Previously, tchaurasiya (Tathagat Chaurasiya) wrote…</i></summary><blockquote>

will this method be used for org/account level environment group?
In that case we should replace this RBAC check.
Also the description,
(Same goes for listEnvironmentsV2() method.
</blockquote></details>

I have taken a note of this and will raise a separate PR for RBAC fixes. 
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/ng/core/infrastructure/services/impl/InfrastructureEntityServiceImpl.java` line 150 at r8](https://reviewable.io/reviews/harness/harness-core/41102#-NJ_oiAd6RutDs8MCQGM:-NJe-9mX9JOGp6tJM6Jf:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/ce140c56027aa89ed14df34b61781a11112631cb/127-cd-nextgen-entities/src/main/java/io/harness/ng/core/infrastructure/services/impl/InfrastructureEntityServiceImpl.java#L150)):*
<details><summary><i>Previously, tchaurasiya (Tathagat Chaurasiya) wrote…</i></summary><blockquote>

should it be envRef?
</blockquote></details>

Done.
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/ng/core/infrastructure/services/impl/InfrastructureEntityServiceImpl.java` line 159 at r8](https://reviewable.io/reviews/harness/harness-core/41102#-NJ_opq8ESe_pq8O3JVW:-NJe-AmW2Ncp4UPsNwZw:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/ce140c56027aa89ed14df34b61781a11112631cb/127-cd-nextgen-entities/src/main/java/io/harness/ng/core/infrastructure/services/impl/InfrastructureEntityServiceImpl.java#L159)):*
<details><summary><i>Previously, tchaurasiya (Tathagat Chaurasiya) wrote…</i></summary><blockquote>

should it be envRef?
</blockquote></details>

Done.
___
*[`127-cd-nextgen-entities/src/main/java/io/harness/ng/core/infrastructure/services/impl/InfrastructureEntityServiceImpl.java` line 258 at r8](https://reviewable.io/reviews/harness/harness-core/41102#-NJ_qZes1n81yPQdUW8S:-NJe-Bqg6O7UD_pPKldu:b-osgqdd) ([raw file](https://github.com/harness/harness-core/blob/ce140c56027aa89ed14df34b61781a11112631cb/127-cd-nextgen-entities/src/main/java/io/harness/ng/core/infrastructure/services/impl/InfrastructureEntityServiceImpl.java#L258)):*
<details><summary><i>Previously, tchaurasiya (Tathagat Chaurasiya) wrote…</i></summary><blockquote>

nit: still needed?
</blockquote></details>

handled scenario where envRef is provided

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 9 of 9 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41102)* status: 35 file(s) to review. 3. unresolved discussions. 1 of 2 approvals obtained.

**Review by tchaurasiya — COMMENTED** [review]
> Reviewed 1 of 5 files at r1, 4 of 19 files at r4, 3 of 11 files at r5, 17 of 25 files at r6, 1 of 1 files at r8, 9 of 9 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41102)* status: 3. unresolved discussions. 1 of 2 approvals obtained.

**Review by tchaurasiya — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41102)* status: :shipit: complete! ✓

**Pull request merged**